### PR TITLE
- add openmpi/4.1.0

### PR DESCRIPTION
**Pull request by SSE4**
based on https://github.com/bincrafters/conan-openmpi
closes: #1858

not in the scope right now:
- Windows builds
- components (cmake/pkg-config)
- dependencies (e.g. libevent)

Specify library name and version:  **openmpi/4.1.0**

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/openmpi/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    url = "https://github.com/conan-io/conan-center-index"
+    topics = ("conan", "mpi", "openmpi")
+    description = "A High Performance Message Passing Library"
+    license = "https://www.open-mpi.org/community/license.php"
+    settings = "os", "arch", "compiler", "build_type"
+    options = {"shared": [True, False],
+               "fPIC": [True, False],
+               "fortran": ['yes', 'mpifh', 'usempi', 'usempi80', 'no']}
+    default_options = {'shared': False, 'fPIC': True, 'fortran': 'no'}
+    _source_subfolder = "sources"
```
Comment (SpaceIm) [inline recipes/openmpi/all/conanfile.py:18]:
```suggestion
    options = {
        "shared": [True, False],
        "fPIC": [True, False],
        "fortran": ['yes', 'mpifh', 'usempi', 'usempi80', 'no']
    }
    default_options = {
        "shared": False,
        "fPIC": True,
        "fortran": "no"
    }

    _autotools = None

    @property
    def _source_subfolder(self):
        return "source_subfolder"
```

File: recipes/openmpi/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    def config(self):
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+        if self.settings.os == "Windows":
+            raise ConanInvalidConfiguration("OpenMPI doesn't support Windows")
```
Comment (SpaceIm) [inline recipes/openmpi/all/conanfile.py:24]:
```suggestion
    def configure(self):
        if self.options.shared:
            del self.options.fPIC
        del self.settings.compiler.libcxx
        del self.settings.compiler.cppstd
        if self.settings.os == "Windows":
            raise ConanInvalidConfiguration("OpenMPI doesn't support Windows")
```

File: recipes/openmpi/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
+from conans import ConanFile, tools, AutoToolsBuildEnvironment
+from conans.errors import ConanInvalidConfiguration
+import os
+
+
+class OpenMPIConan(ConanFile):
```
Comment (SpaceIm) [inline recipes/openmpi/all/conanfile.py:6]:
```suggestion
import os

required_conan_version = ">=1.29.1"


class OpenMPIConan(ConanFile):
```

File: recipes/openmpi/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            args.append('--with-zlib=%s' % self.deps_cpp_info['zlib'].rootpath)
+            args.append('--with-zlib-libdir=%s' % self.deps_cpp_info['zlib'].lib_paths[0])
+            args.append('--datarootdir=${prefix}/res')
+            env_build.configure(args=args)
+            env_build.make()
+            env_build.install()
+
+    def package(self):
+        self.copy(pattern="LICENSE", src='sources', dst='licenses')
+
```
Comment (SpaceIm) [inline recipes/openmpi/all/conanfile.py:56]:
```suggestion
    def _configure_autotools(self):
        if self._autotools:
            return self._autotools
        self._autotools = AutoToolsBuildEnvironment(self)
        args = ["--disable-wrapper-rpath", "--disable-wrapper-runpath"]
        if self.settings.build_type == "Debug":
            args.append("--enable-debug")
        if self.options.shared:
            args.extend(["--enable-shared", "--disable-static"])
        else:
            args.extend(["--enable-static", "--disable-shared"])
        args.append("--with-pic" if self.options.get_safe("fPIC", True) else "--without-pic")
        args.append("--enable-mpi-fortran={}".format(str(self.options.fortran)))
        args.append("--with-zlib={}".format(self.deps_cpp_info["zlib"].rootpath))
        args.append("--with-zlib-libdir={}".format(self.deps_cpp_info["zlib"].lib_paths[0]))
        args.append("--datarootdir=${prefix}/res")
        self._autotools.configure(args=args)
        return self._autotools

    def build(self):
        with tools.chdir(self._source_subfolder):
            autotools = self._configure_autotools()
            autotools.make()

    def package(self):
        self.copy(pattern="LICENSE", src='sources', dst='licenses')
        with tools.chdir(self._source_subfolder):
            autotools = self._configure_autotools()
            autotools.install()
```

File: recipes/openmpi/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
+        tools.rmdir(os.path.join(self.package_folder, "share"))
+        tools.rmdir(os.path.join(self.package_folder, "etc"))
+        tools.remove_files_by_mask(os.path.join(self.package_folder, "lib"), "*.la")
+
+    def package_info(self):
+        self.cpp_info.libs = ['mpi', 'open-rte', 'open-pal']
+        if self.settings.os == "Linux":
+            self.cpp_info.libs.extend(['dl', 'pthread', 'rt', 'util'])
```
Comment (SpaceIm) [inline recipes/openmpi/all/conanfile.py:65]:
```suggestion
            self.cpp_info.system_libs = ["dl", "pthread", "rt", "util"]
```

File: recipes/openmpi/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    def package(self):
+        self.copy(pattern="LICENSE", src='sources', dst='licenses')
+
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
+        tools.rmdir(os.path.join(self.package_folder, "share"))
+        tools.rmdir(os.path.join(self.package_folder, "etc"))
+        tools.remove_files_by_mask(os.path.join(self.package_folder, "lib"), "*.la")
+
+    def package_info(self):
```
Comment (SpaceIm) [inline recipes/openmpi/all/conanfile.py:62]: Is it possible to properly model this module file: https://cmake.org/cmake/help/latest/module/FindMPI.html?

It is called in test_package.

**Comment by SpaceIm** [comment]
closes https://github.com/conan-io/conan-center-index/issues/1858

File: recipes/openmpi/all/conanfile.py (hunk 2 above)
Comment (SpaceIm) [reply recipes/openmpi/all/conanfile.py:24]: By the way, OpenMPI supports windows, but not the recipe, so it's confusing.

File: recipes/openmpi/all/conanfile.py (hunk 3 above)
Comment (SSE4) [reply recipes/openmpi/all/conanfile.py:6]: why?

File: recipes/openmpi/all/conanfile.py (hunk 6 above)
Comment (SSE4) [reply recipes/openmpi/all/conanfile.py:81]: I didn't try, that's too much for the first iteration, as Find module is pretty complex - mpi packages not just libraries, but compiler wrappers

File: recipes/openmpi/all/conanfile.py (hunk 3 above)
Comment (SpaceIm) [reply recipes/openmpi/all/conanfile.py:6]: tools.remove_files_by_mask()

File: recipes/openmpi/all/conanfile.py (hunk 6 above)
Comment (SpaceIm) [reply recipes/openmpi/all/conanfile.py:81]: Therefore could you add a FIXME for CMake and pkg-config?

```
/usr/bin/install -c -m 644 ompi.pc ompi-c.pc ompi-cxx.pc ompi-fort.pc '/home/conan/w/cci_PR-4482/.conan/data/openmpi/4.1.0/_/_/package/b2c1c978ecb1a3aa8058e2e05f1dcb8093dfc54a/lib/pkgconfig'
```

File: recipes/openmpi/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+        if self.settings.os == "Windows":
+            raise ConanInvalidConfiguration("OpenMPI doesn't support Windows")
+
+    def requirements(self):
+        self.requires("zlib/1.2.11")
```
Comment (SpaceIm) [inline recipes/openmpi/all/conanfile.py:40]: openmpi has also a vendored `libevent` (and maybe others libs). Don't know if there is a configure option to use external one.

Comment (SpaceIm) [reply recipes/openmpi/all/conanfile.py:40]: yes you can use external libevent: https://github.com/open-mpi/ompi/blob/master/config/opal_config_libevent.m4

Comment (SSE4) [reply recipes/openmpi/all/conanfile.py:40]: seems to be possible:
```
  --with-libevent=DIR     Search for libevent headers and libraries in DIR.
                          Should only be used if an external copy of libevent
                          is being used.
  --with-libevent-libdir=DIR
                          Search for libevent libraries in DIR. Should only be
                          used if an external copy of libevent is being used.
```
I'll try it

File: recipes/openmpi/all/conanfile.py (hunk 3 above)
Comment (SSE4) [reply recipes/openmpi/all/conanfile.py:6]: okay

**Comment by SSE4** [comment]
for Windows, we had only 3.x in bincrafters which didn't support Windows builds at the time being. I can try Windows build, but I can't promise anything yet

File: recipes/openmpi/all/conanfile.py (hunk 7 above)
Comment (SSE4) [reply recipes/openmpi/all/conanfile.py:42]: weird, it didn't work:
```
checking --with-libevent-libdir value... not found
configure: WARNING: Expected file /Users/sse4/.conan/data/libevent/2.1.12/_/_/package/7ac42c52e63f1381a4948b0f2939b106f29c4ea2/lib/libevent.* not found
```
the files libevent has:
```
➜  all git:(openmpi) ✗ ls /Users/sse4/.conan/data/libevent/2.1.12/_/_/package/7ac42c52e63f1381a4948b0f2939b106f29c4ea2/lib/
libevent_core.a     libevent_extra.a    libevent_openssl.a  libevent_pthreads.a
```

File: recipes/openmpi/all/conanfile.py (hunk 6 above)
Comment (SSE4) [reply recipes/openmpi/all/conanfile.py:83]: what I have right now:
```
[HOOK - conan-center.py] post_package(): ERROR: [PC-FILES (KB-H020)] Found files:
./lib/pkgconfig/ompi-cxx.pc
./lib/pkgconfig/orte.pc
./lib/pkgconfig/ompi-c.pc
./lib/pkgconfig/ompi-f90.pc
./lib/pkgconfig/ompi-f77.pc
./lib/pkgconfig/pmix.pc
./lib/pkgconfig/ompi-fort.pc
./lib/pkgconfig/ompi.pc
```
maybe it's easy to mimic in components, I don't have much experience with it?

File: recipes/openmpi/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+    def package_info(self):
+        self.cpp_info.libs = ['mpi', 'open-rte', 'open-pal']
+        if self.settings.os == "Linux":
+            self.cpp_info.system_libs = ["dl", "pthread", "rt", "util"]
+        self.env_info.MPI_HOME = self.package_folder
+        self.env_info.OPAL_PREFIX = self.package_folder
+        mpi_bin = os.path.join(self.package_folder, 'bin')
+        self.env_info.MPI_BIN = mpi_bin
+        self.env_info.PATH.append(mpi_bin)
```
Comment (SpaceIm) [inline recipes/openmpi/all/conanfile.py:91]:
```suggestion

        self.output.info("Creating MPI_HOME environment variable: {}".format(self.package_folder))
        self.env_info.MPI_HOME = self.package_folder
        self.output.info("Creating OPAL_PREFIX environment variable: {}".format(self.package_folder))
        self.env_info.OPAL_PREFIX = self.package_folder
        mpi_bin = os.path.join(self.package_folder, 'bin')
        self.output.info("Creating MPI_BIN environment variable: {}".format(mpi_bin))
        self.env_info.MPI_BIN = mpi_bin
        self.output.info("Appending PATH environment variable: {}".format(mpi_bin))
        self.env_info.PATH.append(mpi_bin)
```

File: recipes/openmpi/all/test_package/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
+from conans import ConanFile, CMake, tools, RunEnvironment
```
Comment (SpaceIm) [inline recipes/openmpi/all/test_package/conanfile.py:1]:
```suggestion
from conans import ConanFile, CMake, tools
```

File: recipes/openmpi/all/test_package/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        mpiexec = os.path.join(os.environ['MPI_BIN'], 'mpiexec')
+        command = '%s -mca plm_rsh_agent yes -np 2 %s' % (mpiexec, os.path.join("bin", "test_package"))
+        self.run(command, run_environment=True)
```
Comment (SpaceIm) [inline recipes/openmpi/all/test_package/conanfile.py:17]:
```suggestion
    def test(self):
        if not tools.cross_building(self.settings):
            mpiexec = os.path.join(os.environ['MPI_BIN'], 'mpiexec')
            command = '%s -mca plm_rsh_agent yes -np 2 %s' % (mpiexec, os.path.join("bin", "test_package"))
            self.run(command, run_environment=True)
```

File: recipes/openmpi/all/conanfile.py (hunk 6 above)
Comment (SpaceIm) [reply recipes/openmpi/all/conanfile.py:83]: I guess yes. The content of each .pc file would help to see what is the associated lib, and inter dependencies.

File: recipes/openmpi/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+required_conan_version = ">=1.29.1"
+
+
+class OpenMPIConan(ConanFile):
+    name = "openmpi"
+    homepage = "https://www.open-mpi.org"
+    url = "https://github.com/conan-io/conan-center-index"
+    topics = ("conan", "mpi", "openmpi")
+    description = "A High Performance Message Passing Library"
+    license = "https://www.open-mpi.org/community/license.php"
```
Comment (SpaceIm) [inline recipes/openmpi/all/conanfile.py:14]: Is it not `BSD-3-Clause`?

Comment (SSE4) [reply recipes/openmpi/all/conanfile.py:14]: yes, it is

Comment (SSE4) [inline recipes/openmpi/all/conanfile.py:14]:
```suggestion
    license = "BSD-3-Clause"
```

**Review by prince-chrismc — APPROVED** [review]

File: recipes/openmpi/all/conanfile.py (hunk 7 above)
Comment (SpaceIm) [reply recipes/openmpi/all/conanfile.py:42]: @SSE I suggest to add a FIXME to unvendor what we can in another PR, and move forward with this one.

**Review by uilianries — DISMISSED** [review]

**Comment by SpaceIm** [comment]
Please add a FIXME for vendored dependencies, otherwise we'll forget.

**Review by mathbunnyru — CHANGES_REQUESTED** [review]
Only adding FIXME needs to be done here, otherwise good recipe 👍

File: recipes/openmpi/all/conanfile.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+        if self.settings.os == "Windows":
+            raise ConanInvalidConfiguration("OpenMPI doesn't support Windows")
+
+    def requirements(self):
+        self.requires("zlib/1.2.11")
```
Comment (SSE4) [inline recipes/openmpi/all/conanfile.py:42]:
```suggestion
        # FIXME : self.requires("libevent/2.1.12") - try to use libevent from conan
        self.requires("zlib/1.2.11")
```

File: recipes/openmpi/all/conanfile.py (hunk 7 above)
Comment (SSE4) [reply recipes/openmpi/all/conanfile.py]: added

**Review by mathbunnyru — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Comment by SpaceIm** [comment]
I forgot that package_info also requires a FIXME.

**Review by uilianries — APPROVED** [review]

**Pull request merged**